src/core/effects/DepthOfField.vue

```diff
@@ -0,0 +1,121 @@
+<script lang="ts" setup>
+import type { BlendFunction } from 'postprocessing'
+import { EffectPass, DepthOfFieldEffect, BlendMode } from 'postprocessing'
+import { useTresContext } from '@tresjs/core'
+import { inject, onUnmounted, shallowRef, watchEffect } from 'vue'
+import { effectComposerInjectionKey } from '../injectionKeys'
+
+export interface DepthOfFieldProps {
+  /**
+   * The blend function of this effect. This prop is not reactive.
+   */
+  blendFunction?: BlendFunction
+
+  /**
+   * The focus distance in world units.
+   */
+  worldFocusDistance?: number
+
+  /**
+   * The focus range in world units.
+   */
+  worldFocusRange?: number
+
+  /**
+   * The normalized focus distance. Range is [0.0, 1.0].
+   */
+  focusDistance?: number
+
+  /**
+   * The focus range. Range is [0.0, 1.0].
+   */
+  focusRange?: number
+
+  /**
+   * The scale of the bokeh blur.
+   */
+  bokehScale?: number
+
+  resolutionScale?: number
+  resolutionX?: number
+  resolutionY?: number
+}
+
+const props = defineProps<DepthOfFieldProps>()
+
+const composer = inject(effectComposerInjectionKey)
+const pass = shallowRef<EffectPass | null>(null)
+const effect = shallowRef<DepthOfFieldEffect | null>(null)
+
+defineExpose({ pass, effect }) // to allow users to modify pass and effect via template ref
+
+const { camera } = useTresContext()
+
+const unwatch = watchEffect(() => {
+  if (!camera.value || !composer?.value) return
+
+  unwatch?.()
+
+  if (effect.value) return
+
+  effect.value = new DepthOfFieldEffect(camera.value, props)
+  pass.value = new EffectPass(camera.value, effect.value)
+
+  composer?.value?.addPass(pass.value)
+})
+
+watchEffect(() => {
```

Comment:
Question, If you do like this, and a single prop change, the watchEffect will not trigger all the evaluation? Is that what you want, or you just want to "observe" the specific prop that has changed?
Otherwise, could be a possible improvement on performance here, 
In addition, you're creating a const= plainEffectPass ... every single prop change

Reply:
If a single prop changes, all the evaluation will be triggered. Just like you, I'm not 100% satisfied with this code. That's also why I was asking about how we should handle the prop reactivity in Discord. I'm not really sure what the most efficient solution. Here are some thoughts I had:

* the default values of the effect should be set 
   * to preserve maintainability we should not copy the default values from three or postprocessing manually
      * I used the plainEffectPass to get the default values
      * creating an effect in `watchEffect` might not be the most efficient solution. But I also don't want to have an effect in memory that is not used. I'm not sure how to handle this best.
* props should not change too often (see [here](https://tresjs.org/examples/basic-animations.html#but-why-not-using-reactivity))
*  One could create seperate watchEffects for each prop. But I'm not sure whether this is too much overhead.

What do you think about this? Maybe we should have a call about this, because it's a pretty general thing we have to find a common structure/rule for. @JaimeTorrealba @andretchen0 @alvarosabu Wdyt?

Reply:
Well, If you don't need to apply all the changes for every single prop, we could apply watch for every single prop, the other option is to put a bunch of ifs inside the watchEffect

or even computed for single props??

Reply:
@JaimeTorrealba I added some helper functions, that create seperate watchers. A plain effect is only created in case a prop changes to undefined. Wdyt?

Reply:
This is an interesting topic to discuss.

 I'm replacing `watchEffect` with `watch([prop1, prop2])` to respond more granularly to the reactivity.  So in that array, I would place only the props that have an action in common (such as instancing a constructor again), then for others that the action is changing a property of that instance, I use another watcher with an array or individual watchers. 

In general, I see more verbosity on having separate watchers only if it makes sense, and is easier to read. Grouping and automating this too much will lead to unreadability and troubleshooting issues.

I agree with Tino about not using `watchEffect` as a general do-everything-everytime solution, the granular the better.